### PR TITLE
refactor: engine refactor

### PR DESCRIPTION
**Pull request by bsbds**
Depends-On: #703 

Implements `StorageOps` for Engine

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: crates/engine/src/memory_engine/mod.rs [hunk 1]
```diff
@@ -53,6 +54,45 @@ impl MemoryEngine {
...
+                    .ok_or_else(|| EngineError::TableNotFound(table.to_owned()))?;
+                table.retain(|key, _value| {
+                    let key_slice = key.as_slice();
+                    match key_slice.cmp(from) {
+                        Ordering::Less => true,
+                        Ordering::Equal => false,
+                        Ordering::Greater => match key_slice.cmp(to) {
+                            Ordering::Less => false,
+                            Ordering::Equal | Ordering::Greater => true,
+                        },
```
Comment (Phoenix500526) [inline crates/engine/src/memory_engine/mod.rs:89]: nit. Better use `matches!(key_slice.cmp(to), Ordering::Equal | Ordering::Greater)` to replace.

File: crates/engine/src/mock_rocksdb_engine.rs [hunk 2]
```diff
@@ -9,12 +8,12 @@ use bytes::{Bytes, BytesMut};
...
     error::EngineError,
-    memory_engine::{MemoryEngine, MemorySnapshot},
-    TransactionApi, WriteOperation,
+    memory_engine::{MemoryEngine, MemorySnapshot, MemoryTransaction},
+    StorageOps, TransactionApi, WriteOperation,
 };
 
 /// Mock `RocksDB` Storage Engine
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Comment (Phoenix500526) [inline crates/engine/src/mock_rocksdb_engine.rs:16]: Same as above. Why shoule a `RocksEngine` be cloneable?

File: crates/engine/src/mock_rocksdb_engine.rs [hunk 3]
```diff
@@ -206,20 +212,48 @@ impl SnapshotApi for RocksSnapshot {
     }
 }
 
-/// A transaction of the `RocksEngine`
-#[derive(Copy, Clone, Debug, Default)]
+/// Mock `RocksTransaction`
```
Comment (Phoenix500526) [inline crates/engine/src/mock_rocksdb_engine.rs:215]: Shall we rename this `RocksTransaction` as `MockRocksTransaction`? Is it better?

File: crates/engine/src/api/operation.rs [hunk 4]
```diff
@@ -42,7 +43,7 @@ pub trait StorageOps {
 /// Write operation
 #[allow(clippy::module_name_repetitions)]
 #[non_exhaustive]
-#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (bsbds) [reply crates/engine/src/api/operation.rs:44]: The current implementation will write the operation in a loop and needs to be cloned every time. The loop will be removed during the following refactor.

File: crates/engine/src/mock_rocksdb_engine.rs (hunk 2 above)
Comment (bsbds) [reply crates/engine/src/mock_rocksdb_engine.rs:16]: It's a mock implementation, and it's a wrapper around a `MemoryEngine`, which is clonable.

File: crates/engine/src/mock_rocksdb_engine.rs (hunk 3 above)
Comment (bsbds) [reply crates/engine/src/mock_rocksdb_engine.rs:215]: It's meant to replace the normal `RocksTransaction` if the madsim compile flag is set, they should share the same name.

File: crates/engine/src/memory_engine/mod.rs (hunk 1 above)
Comment (bsbds) [reply crates/engine/src/memory_engine/mod.rs]: done

File: crates/engine/src/api/operation.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Phoenix500526) [reply crates/engine/src/api/operation.rs:44]: Shall we leave something, like an issue or comment, to track?

[the remark this answers is not in the document]
Comment (bsbds) [reply crates/engine/src/api/operation.rs:44]: There's a todo comment to track this: https://github.com/xline-kv/Xline/blob/159d74cf02cc63c611788dca2f2b47e8a9f5cbe6/crates/engine/src/rocksdb_engine/mod.rs#L153

**Review by Phoenix500526 — DISMISSED** [review]
LGTM, please squash these commits into one.

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**